x-pack/plugins/actions/server/builtin_action_types/resilient/service.ts

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import axios from 'axios';
+import https from 'https';
+
+import { ExternalServiceCredentials, ExternalService, ExternalServiceParams } from '../case/types';
+import {
+  ResilientPublicConfigurationType,
+  ResilientSecretConfigurationType,
+  CreateIncidentRequest,
+  UpdateIncidentRequest,
+  CreateCommentRequest,
+  UpdateFieldText,
+  UpdateFieldTextArea,
+} from './types';
+
+import * as i18n from './translations';
+import { getErrorMessage, request } from '../case/utils';
+
+const BASE_URL = `rest`;
+const INCIDENT_URL = `incidents`;
+const COMMENT_URL = `comments`;
+
+const VIEW_INCIDENT_URL = `#incidents`;
+
+export const getValueTextContent = (
+  field: string,
+  value: string
+): UpdateFieldText | UpdateFieldTextArea => {
+  if (field === 'description') {
+    return {
+      textarea: {
+        format: 'html',
+        content: value,
+      },
+    };
+  }
+
+  return {
+    text: value,
+  };
+};
+
+export const formatUpdateRequest = ({
+  oldIncident,
+  newIncident,
+}: ExternalServiceParams): UpdateIncidentRequest => {
+  return {
+    changes: Object.keys(newIncident).map((key) => ({
+      field: { name: key },
+      old_value: getValueTextContent(key, oldIncident[key]),
+      new_value: getValueTextContent(key, newIncident[key]),
+    })),
+  };
+};
+
+export const createExternalService = ({
+  config,
+  secrets,
+}: ExternalServiceCredentials): ExternalService => {
+  const { apiUrl: url, orgId } = config as ResilientPublicConfigurationType;
+  const { apiKeyId, apiKeySecret } = secrets as ResilientSecretConfigurationType;
+
+  if (!url || !orgId || !apiKeyId || !apiKeySecret) {
+    throw Error(`[Action]${i18n.NAME}: Wrong configuration.`);
+  }
+
+  const incidentUrl = `${url}/${BASE_URL}/orgs/${orgId}/${INCIDENT_URL}`;
+  const commentUrl = `${incidentUrl}/{inc_id}/${COMMENT_URL}`;
+  const axiosInstance = axios.create({
+    httpsAgent: new https.Agent({
+      rejectUnauthorized: false,
+    }),
+    auth: { username: apiKeyId, password: apiKeySecret },
+  });
+
+  const getIncidentViewURL = (key: string) => {
+    return `${url}/${VIEW_INCIDENT_URL}/${key}`;
+  };
+
+  const getCommentsURL = (incidentId: string) => {
+    return commentUrl.replace('{inc_id}', incidentId);
+  };
+
+  const getIncident = async (id: string) => {
+    try {
+      const res = await request({
+        axios: axiosInstance,
+        url: `${incidentUrl}/${id}?text_content_output_format=objects_convert`,
+      });
```

Comment:
In the current case it is not important, but for the dynamic url params `axios` support setting params as an object, if use it you shouldn't construct the string with `&` and `=` by yourself.